Migration 5 (#104)
neo4j/docs-graphql: modules/ROOT/pages/migration/index.adoc

[[v5-migration]]
:description: This page lists the breaking changes from version 4.0.0 to 5.0.0 and describes how to update.

> :page-aliases: guides/v5-migration/index.adoc , migration/v5-migration/index.adoc

Reviewer: The page-aliases work in a way that, let's say, when a user goes to www.website.com/oldpage, instead of getting a 404 error, they will be redirected to www.website.com/newpage. For that, the alias needs to be added in the newpage. TLDR, we should bring the page-aliases from the old 4.0.0 migration guide here

```suggestion
:page-aliases: guides/index.adoc, guides/migration-guide/index.adoc, guides/migration-guide/server.adoc, guides/migration-guide/type-definitions.adoc, guides/migration-guide/mutations.adoc
```

== Breaking changes

Here is a list of all the breaking changes from version 4.0.0 to 5.0.0.

=== `@relationshipProperties`

> The directive `@relationshipProperties` should now be used on types rather than interfaces:

Reviewer: ```suggestion
The directive `@relationshipProperties` should now be used on types rather than interfaces.
For example:
```

[cols="1,1"]
|===
|Before | Now

a|
[source, graphql, indent=0]
----
interface ActedIn @relationshipProperties {
    screenTime: Int
}

type Movie {
    title: String
    actors: [Person!]! @relationship(type: "ACTED_IN", direction: IN, properties: "ActedIn")
}

type Person {
    name: String
}
----
a|
[source, graphql, indent=0]
----
type ActedIn @relationshipProperties {
    screenTime: Int
}

type Movie {
    title: String
    actors: [Person!]! @relationship(type: "ACTED_IN", direction: IN, properties: "ActedIn")
}

type Person {
    name: String
}
----
|===

=== Interface directives

> To better match the default behavior of GraphQL, library directives used in interfaces will not cascade to the implementing types anymore. This means that most directives are no longer valid in interfaces and have to be defined in the implementing types. This also applies to custom directives.

Reviewer: I understand that this is more general, but is it possible to add a code example just to clarify that for users?

```suggestion
To better match the default behavior of GraphQL, library directives used in interfaces will not cascade to the implementing types anymore. 
This means that most directives are no longer valid in interfaces and have to be defined in the implementing types. 
This also applies to custom directives.
```

=== `@declareRelationship`

> The `@relationship` directive is no longer available in interfaces. If you need a relationship to be available on interfaces, you'll need to use the new `@declareRelationship` directive instead, as well as define the relationships in the concrete type.

Reviewer: ```suggestion
The `@relationship` directive is no longer available in interfaces. 
If you need a relationship to be available on interfaces, you need to use the new `@declareRelationship` directive instead, as well as define the relationships in the concrete type.
```

> This is due to directives no longer cascading from interfaces to types and allows for a relationship to have different properties and labels in each type.

Reviewer: ```suggestion
This change is due to directives no longer cascading from interfaces to types.
However, now it is possible for a relationship to have different properties and labels in each type.
For example:
```

}

type Movie implements Production {
    title: String!
    actors: [Actor!]!
}

type Series implements Production {
    title: String!
    episodes: Number!
    actors: [Actor!]! 
}

type Actor {
    name: String!
}
----
a|
[source, graphql, indent=0]
----
interface Production {
    title: String!
    actors: [Actor!]! @declareRelationship
}

type Movie implements Production {
    title: String!
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)
}

type Series implements Production {
    title: String!
    episodes: Number!
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)
}

type Actor {
    name: String!
}
----
|===

=== Edge properties

Edge properties in connections now exist within the field `properties` in `edges`:

[cols="1,1"]
|===
|Before | Now
a|
[source, graphql, indent=0]
----
query ActedInConnection {
  actors {
    actedInConnection {
      edges {
          screenTime
      }
    }
  }
}
----
a|
[source, graphql, indent=0]
----
query ActedInConnection {
  actors {
    actedInConnection {
      edges {
        properties {
          screenTime
        }
      }
    }
  }
}
----
|===

> === `_on` filters deprecated

Reviewer: ```suggestion
=== Deprecations
```

> `_on` filters for interfaces are no longer available in `where` and mutations. To filter by a implementing type you need to use the new filter `typename_IN`: 

Reviewer: ```suggestion
`_on` filters for interfaces are no longer available in `where` and mutations. 
To filter by a implementing type, you need to use the new filter `typename_IN`: 
```

> `_on` filters for interfaces are no longer available in `where` and mutations. To filter by a implementing type you need to use the new filter `typename_IN`: 

Reviewer: ```suggestion
==== `_on` filters

`_on` filters for interfaces are no longer available in `where` and mutations. To filter by a implementing type you need to use the new filter `typename_IN`: 
```

[cols="1,1"]
|===
|Before | Now
a|
[source, graphql, indent=0]
----
query MyQuery {
  actors(
    where: {
      actedInConnection_SINGLE: { node: { _on: { Movie: { } } } }
    }
  ) {
    name
  }
}
----
a|
[source, graphql, indent=0]
----
query MyQuery {
 actors(
    where: {
      actedInConnection_SINGLE: { node: { typename_IN: [Movie] } }
    }
  ) {
    name
  }
}
----
|===

[NOTE]
====
Using fields of a type in an interface operation is no longer supported.
====

> === Minimum Node updated to `18.0.0`

Reviewer: ```suggestion
==== Minimum NodeJS version
```

> With the deprecation of Node 16, the minimum supported NodeJS version is `18.0.0`

Reviewer: ```suggestion
With the deprecation of Node 16, the minimum supported NodeJS version is `18.0.0`.
```

> === `experimental` flag removed

Reviewer: ```suggestion
==== `experimental` flag
```

> The `experimental` flag is no longer available on the options of `Neo4jGraphQL` class.

Reviewer: ```suggestion
The `experimental` flag is no longer available in the options of the `Neo4jGraphQL` class.
```
